### DISCUSSION TITLE
Making Language Server Protocol work

### DISCUSSION THREAD

**Discussion by juergen66** [jupyterlab/jupyterlab#15145] [Q&A]
I installed a fresh virtual environment with JupyterLab 4.0.6 and pylsp:
```
python3 -m venv venv
. venv/bin/activate
pip3 install jupyterlab
pip3 install 'python-lsp-server[all]'

```
My LSP configuration looks like this:
```
{
 
   "activate": "on",
   "languageServers": {
        "pylsp": {
            "configuration": {
                "pylsp.plugins.flake8.enabled": true
            },
            "rank": 50
        }
    },
    "logAllCommunication": true,
    "setTrace": "off"
}
```

Still, nothing seems to happen. I can make as many errors as I want - nothing gets highlighted.

Sometimes, when I Ctrl-C JupyterLab, I get a stack trace:
```
  File "/home/juergen/JupyterLab/venv/lib/python3.11/site-packages/pylsp/python_lsp.py", line 206, in start
    self._jsonrpc_stream_reader.listen(self._endpoint.consume)
  File "/home/juergen/JupyterLab/venv/lib/python3.11/site-packages/pylsp_jsonrpc/streams.py", line 30, in listen
    request_str = self._read_message()
                  ^^^^^^^^^^^^^^^^^^^^
  File "/home/juergen/JupyterLab/venv/lib/python3.11/site-packages/pylsp_jsonrpc/streams.py", line 51, in _read_message
    line = self._rfile.readline()
           ^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt
```
I guess that means that the LSP client is listening?

Does anyone have LSP working in JupyterLab 4.0.x, and provide some hints?

Thanks.

**Comment by krassowski** [comment]
JupyterLab 4.0 does not ship with any LSP features enabled (maybe this should be clarified in the documentation). To get LSP integration you can install [`jupyterlab-lsp`](https://github.com/jupyter-lsp/jupyterlab-lsp) package in version 5.0 (as of day of writing in in release candidate phase: [v5.0.0rc0](https://github.com/jupyter-lsp/jupyterlab-lsp/releases/tag/v5.0.0rc0)).

**Reply by hutch3232** [reply]
Could you clarify if `jupyterlab-lsp` is still required? As alluded to by @juergen66, the docs say:

> JupyterLab 4.0 ships with components on both backend and frontend to lay the infrastructure for handling and communicating with language servers.
> 
> This means that other extensions can be built on top of JupyterLab’s services to provide LSP features for notebooks and files.
> 
> The LSP service is disabled by default, it can be enabled by togging the Activate option in the Language Servers section of the setting page.

https://jupyterlab.readthedocs.io/en/latest/user/lsp.html

My read of that is that `jupyterlab-lsp` is essentially baked into `jupyterlab` now and doesn't need to be installed separately. All that is needed is to enable it in the settings and install a compatible lsp.

My experience though is that the installed lsp does not appear to work - errors are not highlighted.

**Reply by krassowski** [reply]
> My read of that is that `jupyterlab-lsp` is essentially baked into jupyterlab now and doesn't need to be installed separately. All that is needed is to enable it in the settings and install a compatible lsp.

Unfortunately, no. `jupyterlab-lsp` is not baked into jupyterlab, only some parts were cherry picked and baked into the jupyterlab; which caused me a bit of trouble porting `jupyterlab-lsp` to support lab 4+.

**Comment by juergen66** [comment]
Bummer. https://jupyterlab.readthedocs.io/en/latest/user/lsp.html seemed to
say that it was available.

I guess I'll give jupyterlab-lsp another try.

Thanks for letting me know.

**Comment by krassowski** [comment]
Here you go @juergen66 and @hutch3232, does the rewritten docs as in https://github.com/jupyterlab/jupyterlab/pull/16160 improve the situation?

Of note, there is more talk about whether the `jupyterlab-lsp` extension should be moved to the `jupyterlab` org on the governance side (https://github.com/jupyter-governance/ec-team-compass/issues/25), I don't think it changes much as any major changes on the code side of it will not happen unless someone volunteers to help (or to sponsor it).

**Reply by hutch3232** [reply]
Thank you, it's very clear to me now. LSP is such a useful feature, the easier to get started/more integrated, the better in my opinion. Appreciate your efforts!
